docarray/storage/backends/HnswDocStore.py

```diff
@@ -0,0 +1,352 @@
+import hashlib
+import os
+import sqlite3
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Any, Dict, Generic, List, Sequence, Tuple, Type, TypeVar, Union
+
+import hnswlib
+import numpy as np
+
+import docarray.typing
+from docarray import BaseDocument, DocumentArray
+from docarray.proto import DocumentProto
+from docarray.storage.abstract_doc_store import (
+    BaseDocumentStore,
+    FindResultBatched,
+    _Column,
+    composable,
+)
+from docarray.typing import AnyTensor
+from docarray.utils.filter import filter as da_filter
+from docarray.utils.find import FindResult
+from docarray.utils.misc import torch_imported
+
+TSchema = TypeVar('TSchema', bound=BaseDocument)
+
+HNSWLIB_PY_VEC_TYPES = [list, tuple, np.ndarray]
+if torch_imported:
+    import torch
+
+    HNSWLIB_PY_VEC_TYPES.append(torch.Tensor)
+
+
+class HnswDocumentStore(BaseDocumentStore, Generic[TSchema]):
+    def __init__(self, db_config=None, **kwargs):
+        super().__init__(db_config=db_config, **kwargs)
+        self._work_dir = self._db_config.work_dir
+        load_existing = os.path.exists(self._work_dir)
+        Path(self._work_dir).mkdir(parents=True, exist_ok=True)
+
+        # HNSWLib setup
+        self._index_construct_params = ('space', 'dim')
+        self._index_init_params = (
+            'max_elements',
+            'ef_construction',
+            'M',
+            'allow_replace_deleted',
+        )
+
+        self._hnsw_locations = {
+            col_name: os.path.join(self._work_dir, f'{col_name}.bin')
+            for col_name, col in self._columns.items()
+            if col.config
+        }
+        self._hnsw_indices = {}
+        for col_name, col in self._columns.items():
+            if not col.config:
+                continue  # do not create column index if no config is given
+            if load_existing:
+                self._hnsw_indices[col_name] = self._load_index(col_name, col)
+            else:
+                self._hnsw_indices[col_name] = self._create_index(col)
+
+        # SQLite setup
+        self._sqlite_db_path = os.path.join(self._work_dir, 'docs_sqlite.db')
+        self._sqlite_conn = sqlite3.connect(self._sqlite_db_path)
+        self._sqlite_cursor = self._sqlite_conn.cursor()
+        self._create_docs_table()
+        self._sqlite_conn.commit()
+
+    ###############################################
+    # Inner classes for query builder and configs #
+    ###############################################
+    class QueryBuilder(BaseDocumentStore.QueryBuilder):
+        def build(self, *args, **kwargs) -> Any:
+            return self._queries
+
+    @dataclass
+    class DBConfig(BaseDocumentStore.DBConfig):
+        work_dir: str = '.'
+
+    @dataclass
+    class RuntimeConfig(BaseDocumentStore.RuntimeConfig):
```

Comment:
it is not clear to me how and where does people provide this.

Reply:
what do you mean by "people"?
Backend devs put the dataclass as inner class in their implementation, and users can pass it to the `configure` method